staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -94,6 +93,16 @@ type Controller struct {
 	nodeListerSynced    cache.InformerSynced
 	// services that need to be synced
 	queue workqueue.RateLimitingInterface
+
+	// triggerLock ensures there is only one instance of triggerNodeSync getting executed at one time
+	triggerLock sync.Mutex
+	// nodeSyncCh triggers nodeSyncLoop to run
+	nodeSyncCh chan interface{}
+
+	// needFullSync indicates if the nodeSyncInternal will do a full node sync on all LB services.
+	needFullSync bool
+	// fullSyncLock is the lock for needFullSync value
+	fullSyncLock sync.Mutex
```

Comment:
I know that technically having two mutexes might help a little bit with contention here, but:
- the contention shouldn't be visible
- it's relatively easy to introduce deadlock with two locks
- the "fullSyncLock" you're always using when the triggerLock is actually help (so it only helps in one place)

I would suggest to sticking to a single lock (and protect both the compute part of newNodes and needFullSync field with it).

Reply:
Ack. I had the same concern before I write this. I would convert this to one lock. 

But just keep in mind that the functions that operate on `needFullSync` becomes asymmetric.